lib/src/common/cache/cache.dart

```diff
@@ -0,0 +1,300 @@
+  /// [CachingStrategy.onDidGet] method or [valueFactory].
+  ///
+  ///     var a = cache.getAsync('id', _superLongAsyncCall);
+  ///     var b = cache.getAsync('id', _superLongAsyncCall);
+  ///     var c = cache.getAsync('id', _superLongAsyncCall);
+  ///
+  ///     // The futures in this list will all resolve to the same instance,
+  ///     // which was returned by `_superLongAsyncCall` the first time `get` was
+  ///     // called.
+  ///     var values = Future.wait([a, b, c]);
+  ///
+  /// If the [Cache] [isOrWillBeDisposed] then a [StateError] is thrown.
+  @mustCallSuper
+  Future<TValue> getAsync(TIdentifier id, Func<Future<TValue>> valueFactory) {
+    _throwWhenDisposed('getAsync');
+    _cachingStrategy.onWillGet(id);
+    // Await any pending cached futures
+    if (_cache.containsKey(id)) {
+      return _cache[id].then((TValue value) async {
+        await _cachingStrategy.onDidGet(id, value);
+        return value;
+      });
+    }
+
+    // Install Future value
+    _cache[id] = valueFactory.call().then((TValue value) async {
+      await _cachingStrategy.onDidGet(id, value);
+      _didUpdateController.add(new CacheContext(id, value));
+      return value;
+    });
+
+    return _cache[id];
+  }
+
+  Iterable<TIdentifier> get keys => _cache.keys;
+
+  /// Releases a reference for a given [TIdentifier] and removes the
+  /// [TIdentifier] from the [Cache] when the last reference is released.
+  ///
+  /// If [Cache] [isOrWillBeDisposed] then a [StateError] is
+  /// thrown.
+  @mustCallSuper
+  Future<Null> release(TIdentifier id) {
+    _throwWhenDisposed('release');
+    // Await any pending cached futures
+    if (_cache.containsKey(id)) {
+      _cachingStrategy.onWillRelease(id);
+      return _cache[id].then((TValue value) async {
+        await _cachingStrategy.onDidRelease(id, value, remove);
+        _didReleaseController.add(new CacheContext(id, value));
+      }).catchError((Object error, StackTrace stackTrace) {
+        return null;
+      });
+    }
+
+    return new Future.value();
+  }
+
+  /// Removes the reference to a [TValue] associated with the given
+  /// [TIdentifier].
+  ///
+  /// A [CacheContext] value is emitted by the [didUpdate] stream upon
+  /// successful removal of the given [TIdentifier]. In addition, a
+  /// [CacheContext] value is emitted by the [didRemove] stream upon successful
+  /// removal from the cache.
+  ///
+  /// If the [Cache] [isOrWillBeDisposed] then a [StateError] is thrown.
+  @mustCallSuper
+  Future<Null> remove(TIdentifier id) {
+    _throwWhenDisposed('remove');
+    if (_cache.containsKey(id)) {
+      final removedValue = _cache.remove(id);
+      _cachingStrategy.onWillRemove(id);
+      return removedValue.then((TValue value) async {
+        await _cachingStrategy.onDidRemove(id, value);
+        _didRemoveController.add(new CacheContext(id, value));
+        _didUpdateController.add(new CacheContext(id, null));
```

Comment:
Ah, I see we're using didUpdate here too. So maybe we don't want to change it to didAdd.